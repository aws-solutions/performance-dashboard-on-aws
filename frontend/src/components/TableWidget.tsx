/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { useMemo, useState } from "react";
import { ColumnMetadata } from "../models";
import { useTableMetadata } from "../hooks";
import MarkdownRender from "./MarkdownRender";
import TickFormatter from "../services/TickFormatter";
import UtilsService from "../services/UtilsService";
import Table from "./Table";
import ShareButton from "./ShareButton";

type Props = {
    id: string;
    title: string;
    summary: string;
    data?: Array<object>;
    summaryBelow: boolean;
    columnsMetadata: ColumnMetadata[];
    sortByColumn?: string;
    sortByDesc?: boolean;
    significantDigitLabels: boolean;
    displayWithPages: boolean;
    showMobilePreview?: boolean;
};

const TableWidget = ({
    id,
    data,
    summaryBelow,
    summary,
    title,
    columnsMetadata,
    sortByDesc,
    sortByColumn,
    significantDigitLabels,
    displayWithPages,
}: Props) => {
    const { largestTickByColumn } = useTableMetadata(data);
    const [filteredJson, setFilteredJson] = useState<any[]>([]);

    useMemo(() => {
        let headers = data && data.length > 0 ? Object.keys(data[0]) : [];
        headers = headers.filter((h) => {
            const metadata = columnsMetadata
                ? columnsMetadata.find((c) => c.columnName === h)
                : undefined;
            return !metadata || !metadata.hidden;
        });
        const newFilteredJson = new Array<any>();
        for (const row of data || new Array<any>()) {
            const filteredRow = headers.reduce((obj: any, key: any) => {
                obj[key] = row[key];
                return obj;
            }, {});
            if (Object.keys(filteredRow).length > 0) {
                newFilteredJson.push(filteredRow);
            }
        }
        setFilteredJson(newFilteredJson);
    }, [data, columnsMetadata]);

    const rows = useMemo(() => filteredJson || [], [filteredJson]);
    const columns = useMemo(() => {
        const keys = filteredJson.length ? Object.keys(filteredJson[0] as Array<string>) : [];

        return keys.map((header) => {
            return {
                Header: header,
                id: header,
                accessor: header,
                minWidth: 150,
                Cell: (props: any) => {
                    const row = props.row.original;

                    // Check if there is metadata for this column
                    let columnMetadata;
                    if (columnsMetadata) {
                        columnMetadata = columnsMetadata.find((cm) => cm.columnName === header);
                    }

                    return !UtilsService.isCellEmpty(row[header])
                        ? TickFormatter.format(
                              row[header],
                              largestTickByColumn[header],
                              significantDigitLabels,
                              "",
                              "",
                              columnMetadata,
                          )
                        : "-";
                },
            };
        });
    }, [largestTickByColumn, significantDigitLabels, columnsMetadata, filteredJson]);

    if (!filteredJson || filteredJson.length === 0) {
        return null;
    }

    return (
        <div
            aria-label={title}
            tabIndex={-1}
            className={`overflow-x-hidden overflow-y-hidden ${title ? "" : "padding-top-2"}`}
        >
            {title && (
                <h2 className={`margin-bottom-${summaryBelow ? "4" : "1"}`}>
                    {title}
                    <ShareButton id={`${id}a`} title={title} className="margin-left-1" />
                </h2>
            )}
            {!summaryBelow && (
                <MarkdownRender
                    source={summary}
                    className="usa-prose margin-top-0 margin-bottom-3 tableSummaryAbove textOrSummary"
                />
            )}
            <Table
                id={id}
                selection="none"
                rows={rows}
                initialSortAscending={sortByDesc !== undefined ? !sortByDesc : true}
                initialSortByField={sortByColumn}
                disablePagination={!displayWithPages && rows.length < 25}
                columns={columns}
                sortByColumn={sortByColumn}
                sortByDesc={sortByDesc}
                title={title}
                mobileNavigation={columns.length < 3}
                pageSize={25}
            />
            {summaryBelow && (
                <MarkdownRender
                    source={summary}
                    className="usa-prose margin-top-3 margin-bottom-0 tableSummaryBelow textOrSummary"
                />
            )}
        </div>
    );
};

export default TableWidget;
