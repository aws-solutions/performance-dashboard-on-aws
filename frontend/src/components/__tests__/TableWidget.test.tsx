/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import TableWidget from "../TableWidget";

import { library } from "@fortawesome/fontawesome-svg-core";
import { faCaretUp, faCaretDown } from "@fortawesome/free-solid-svg-icons";

library.add(faCaretUp, faCaretDown);

test("renders the table title", async () => {
    render(
        <TableWidget
            title="test title"
            summary="test summary"
            data={[
                {
                    id: "1",
                    name: "Banana",
                    updatedAt: "2021-11-11",
                },
                {
                    id: "2",
                    name: "Chocolate",
                    updatedAt: "2020-11-11",
                },
                {
                    id: "3",
                    name: "Vanilla",
                    updatedAt: "2019-11-11",
                },
            ]}
            summaryBelow={false}
            columnsMetadata={[]}
            significantDigitLabels={false}
            displayWithPages={false}
        />,
        { wrapper: MemoryRouter },
    );

    expect(screen.getByText("test title")).toBeInTheDocument();

    const summary = screen.getByText("test summary");
    expect(summary).toBeInTheDocument();
    expect(summary.closest("div")).toHaveClass("tableSummaryAbove");
});

test("renders the summary below the chart", async () => {
    render(
        <TableWidget
            title="test title"
            summary="test summary"
            data={[
                {
                    id: "1",
                    name: "Banana",
                    updatedAt: "2021-11-11",
                },
                {
                    id: "2",
                    name: "Chocolate",
                    updatedAt: "2020-11-11",
                },
                {
                    id: "3",
                    name: "Vanilla",
                    updatedAt: "2019-11-11",
                },
            ]}
            summaryBelow={true}
            columnsMetadata={[]}
            significantDigitLabels={false}
            displayWithPages={false}
        />,
        { wrapper: MemoryRouter },
    );
    const summary = screen.getByText("test summary");
    expect(summary).toBeInTheDocument();
    expect(summary.closest("div")).toHaveClass("tableSummaryBelow");
});

test("table preview should match snapshot", async () => {
    const wrapper = render(
        <TableWidget
            title="test title"
            summary="test summary"
            data={[
                {
                    id: "1",
                    name: "Banana",
                    updatedAt: "2021-11-11",
                },
                {
                    id: "2",
                    name: "Chocolate",
                    updatedAt: "2020-11-11",
                },
                {
                    id: "3",
                    name: "Vanilla",
                    updatedAt: "2019-11-11",
                },
            ]}
            summaryBelow={false}
            columnsMetadata={[]}
            significantDigitLabels={false}
            displayWithPages={true}
        />,
        { wrapper: MemoryRouter },
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("table should not crash when a column header is an empty string", async () => {
    render(
        <TableWidget
            title="test title"
            summary="test summary"
            data={[
                {
                    id: "",
                    name: "Banana",
                    updatedAt: "2021-11-11",
                },
                {
                    id: "2",
                    name: "Chocolate",
                    updatedAt: "2020-11-11",
                },
                {
                    id: "3",
                    name: "Vanilla",
                    updatedAt: "2019-11-11",
                },
            ]}
            summaryBelow={false}
            columnsMetadata={[]}
            significantDigitLabels={false}
            displayWithPages={false}
        />,
        { wrapper: MemoryRouter },
    );

    expect(screen.getByRole("table")).toBeInTheDocument();
    expect(screen.getByText("Banana")).toBeInTheDocument();
    expect(screen.getByText("Chocolate")).toBeInTheDocument();
    expect(screen.getByText("Vanilla")).toBeInTheDocument();
});
