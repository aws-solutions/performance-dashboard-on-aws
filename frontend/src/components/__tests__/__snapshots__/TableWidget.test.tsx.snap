// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table preview should match snapshot 1`] = `
<div>
  <div
    aria-label="test title"
    class="overflow-x-hidden overflow-y-hidden "
    tabindex="-1"
  >
    <h2
      class="margin-bottom-1"
    >
      test title
      <a
        aria-label="Copy link: test title"
        class="shareButton margin-left-1"
        href="http://localhost/#undefineda"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-link fa-w-16 fa-xs "
          data-icon="link"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
            fill="currentColor"
          />
        </svg>
      </a>
      <span
        class="anchor"
        id="undefineda"
      />
    </h2>
    <div
      class="Markdown usa-prose margin-top-0 margin-bottom-3 tableSummaryAbove textOrSummary"
    >
      <p>
        test summary
      </p>
    </div>
    <div
      class="overflow-x-hidden overflow-y-hidden"
    >
      <table
        class="usa-table usa-table--borderless"
        role="table"
      >
        <thead>
          <tr
            role="row"
          >
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 150px;"
            >
              <button
                aria-label="Toggle SortBy: id"
                class="usa-button usa-button--unstyled text-no-underline"
                style="cursor: pointer;"
                title="Toggle SortBy: id"
                type="button"
              >
                <span>
                  id
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 150px;"
            >
              <button
                aria-label="Toggle SortBy: name"
                class="usa-button usa-button--unstyled text-no-underline"
                style="cursor: pointer;"
                title="Toggle SortBy: name"
                type="button"
              >
                <span>
                  name
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 150px;"
            >
              <button
                aria-label="Toggle SortBy: updatedAt"
                class="usa-button usa-button--unstyled text-no-underline"
                style="cursor: pointer;"
                title="Toggle SortBy: updatedAt"
                type="button"
              >
                <span>
                  updatedAt
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <tr
            role="row"
          >
            <th
              role="cell"
              scope="row"
            >
              1
            </th>
            <td
              role="cell"
            >
              Banana
            </td>
            <td
              role="cell"
            >
              2021-11-11
            </td>
          </tr>
          <tr
            role="row"
          >
            <th
              role="cell"
              scope="row"
            >
              2
            </th>
            <td
              role="cell"
            >
              Chocolate
            </td>
            <td
              role="cell"
            >
              2020-11-11
            </td>
          </tr>
          <tr
            role="row"
          >
            <th
              role="cell"
              scope="row"
            >
              3
            </th>
            <td
              role="cell"
            >
              Vanilla
            </td>
            <td
              role="cell"
            >
              2019-11-11
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="grid-row text-base-darker text-italic padding-y-05 padding-right-1"
      >
        <div
          class="grid-col-6 text-left"
        >
          <div
            style="display: inline-flex;"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-download fa-w-16 fa-xs margin-right-1"
              data-icon="download"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M216 0h80c13.3 0 24 10.7 24 24v168h87.7c17.8 0 26.7 21.5 14.1 34.1L269.7 378.3c-7.5 7.5-19.8 7.5-27.3 0L90.1 226.1c-12.6-12.6-3.7-34.1 14.1-34.1H192V24c0-13.3 10.7-24 24-24zm296 376v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h146.7l49 49c20.1 20.1 52.5 20.1 72.6 0l49-49H488c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            style="display: inline-flex;"
          >
            <div
              class="margin-right-05"
            >
              <a
                class="text-base-darker"
                download="test title"
                href="data:text/csv;charset=utf-8,ï»¿\\"id\\",\\"name\\",\\"updatedAt\\"
\\"1\\",\\"Banana\\",\\"2021-11-11\\"
\\"2\\",\\"Chocolate\\",\\"2020-11-11\\"
\\"3\\",\\"Vanilla\\",\\"2019-11-11\\""
                target="_self"
              >
                Download table
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
